File: leetcode/15.3Sum.go
package leetcode

import "sort"

func threeSum(nums []int) [][]int {
	var result [][]int

	sort.Ints(nums)

	right := 0
	left := 0

	for i := 0; i < len(nums); i++ {
		num := nums[i]

		if i-1 < len(nums) && i-1 >= 0 && num == nums[i-1] {
			continue
		}
		if num > 0 {
			break
		}

		left = i + 1
		right = len(nums) - 1

		for left < right {
			sum := num + nums[left] + nums[right]
			if sum > 0 {
				right--
				continue
			}
			if sum < 0 {
				left++
				continue
			}

			result = append(result, [][]int{
				{num, nums[left], nums[right]},
			}...)

			left++
			for nums[left] == nums[left-1] && left < right {
				left++
			}
		}
	}

	return result
}

File: leetcode/15.3Sum.ts
function threeSum(nums: number[]): number[][] {
  const result = []
  nums.sort((a, b) => a - b)

  let right = 0
  let left = 0

  for (let i = 0; i < nums.length; i++) {
    const num = nums[i]

    if (num === nums[i - 1]) continue
    if (num > 0) break

    left = i + 1
    right = nums.length - 1
    while (left < right) {
      const sum = num + nums[left] + nums[right]
      if (sum > 0) {
        right--
        continue
      }
      if (sum < 0) {
        left++
        continue
      }

      result.push([num, nums[left], nums[right]])

      left++
      while (nums[left] === nums[left - 1] && left < right) {
        left++
      }
    }
  }

  return result
}
console.log(threeSum([-1, 0, 1, 2, -1, -4]))
